hash_test.go

```diff
@@ -0,0 +1,140 @@
+package id_test
+
+import (
+	"crypto/rand"
+	"testing"
+	"testing/quick"
+
+	"github.com/renproject/id"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Hashes", func() {
+	Context("when computing the merkle hash", func() {
+		Context("when using the safe and unsafe implementations", func() {
+			Context("when computing the merkle hash of zero hashes", func() {
+				It("should return the empty hash", func() {
+					f := func() bool {
+						hashes := make([]id.Hash, 0)
+						rootHash := id.NewMerkleHash(hashes)
+						safeRootHash := id.NewMerkleHashSafe(hashes)
+						Expect(rootHash).To(Equal(id.Hash{}))
+						Expect(safeRootHash).To(Equal(id.Hash{}))
+						return true
+					}
+					Expect(quick.Check(f, nil)).To(Succeed())
+				})
+			})
+
+			Context("when computing the merkle hash of one hash", func() {
+				It("should return the input hash", func() {
+					f := func() bool {
+						hashes := make([]id.Hash, 1)
+						for i := range hashes {
+							rand.Read(hashes[i][:])
+						}
+						rootHash := id.NewMerkleHash(hashes)
+						safeRootHash := id.NewMerkleHashSafe(hashes)
+						Expect(rootHash).To(Equal(hashes[0]))
+						Expect(safeRootHash).To(Equal(hashes[0]))
+						return true
+					}
+					Expect(quick.Check(f, nil)).To(Succeed())
+				})
+			})
+
+			Context("when computing the merkle hash of two hashes", func() {
+				It("should return the same merkle hash", func() {
+					f := func() bool {
+						hashes := make([]id.Hash, 2)
+						for i := range hashes {
+							rand.Read(hashes[i][:])
+						}
+						rootHash := id.NewMerkleHash(hashes)
+						safeRootHash := id.NewMerkleHashSafe(hashes)
+						Expect(rootHash).To(Equal(safeRootHash))
+						return true
+					}
+					Expect(quick.Check(f, nil)).To(Succeed())
+				})
+			})
+
+			Context("when computing the merkle hash of three hash", func() {
+				It("should return the same merkle hash", func() {
+					f := func() bool {
+						hashes := make([]id.Hash, 3)
+						for i := range hashes {
+							rand.Read(hashes[i][:])
+						}
+						rootHash := id.NewMerkleHash(hashes)
+						safeRootHash := id.NewMerkleHashSafe(hashes)
+						Expect(rootHash).To(Equal(safeRootHash))
```

Comment:
I think the reason that the above bugs are not caught is that here we only test against the two implementations. I believe we should also compare the results of each implementation to manually computing the hash.

Also, what is the standard for using quick check; I was expecting that instead of using a `gomega` checker that you would return `rootHash == safeRootHash` instead.